tensorflow_probability/python/mcmc/transformed_kernel_test.py

```diff
@@ -308,6 +326,89 @@ def test_is_calibrated(self):
+    new_kernel = step_then_transform(tfb.Exp())
+    pkr_one, pkr_two = self.evaluate([
+        new_kernel.bootstrap_results(2.),
+        new_kernel.bootstrap_results(9.),
+    ])
+    self.assertNear(np.log(2.), pkr_one.transformed_state, err=1e-6)
+    self.assertNear(np.log(9.), pkr_two.transformed_state, err=1e-6)
+
+  def test_step_size_changed(self):
+    target_dist = tfd.MultivariateNormalDiag(loc=[0., 0.], scale_diag=[1., 10.])
+    # `hmc_kernel`'s step size is far from optimal
+    hmc_kernel = tfp.mcmc.HamiltonianMonteCarlo(
+        target_log_prob_fn=target_dist.log_prob,
+        num_leapfrog_steps=27,
+        step_size=10,
+        seed=_maybe_seed(test_util.test_seed()))
+    step_adaptation_kernel = tfp.mcmc.SimpleStepSizeAdaptation(
+        inner_kernel=hmc_kernel,
+        adaptation_rate=0.8,
+        num_adaptation_steps=9)
+    trans_kernel = tfp.mcmc.TransformedTransitionKernel(
+        inner_kernel=step_adaptation_kernel,
+        bijector=tfb.Exp()
+    )
+    kernel_results = trans_kernel.inner_kernel.bootstrap_results(tf.zeros(2))
+    for _ in range(2):
+      _, kernel_results = trans_kernel.inner_kernel.one_step(tf.zeros(2),
+                                                             kernel_results)
+    adapted_step_size = self.evaluate(
+        kernel_results.inner_results.accepted_results.step_size)
+    self.assertLess(adapted_step_size, 7)
+
+  def test_deeply_nested(self):
+    trans_kernel = tfp.mcmc.TransformedTransitionKernel(
+        inner_kernel=FakeInnerKernel(target_log_prob_fn=fake_target_log_prob),
+        bijector=tfb.Exp())
+    step_kernel = tfp.mcmc.SimpleStepSizeAdaptation(
+        inner_kernel=trans_kernel,
+        num_adaptation_steps=9)
+    double_step_kernel = tfp.mcmc.SimpleStepSizeAdaptation(
+        inner_kernel=step_kernel,
+        num_adaptation_steps=9)
+    pkr_one, pkr_two = self.evaluate([
+        double_step_kernel.bootstrap_results(2.),
+        double_step_kernel.bootstrap_results(9.),
+    ])
+    self.assertNear(np.log(2.),
+                    pkr_one.inner_results.inner_results.transformed_state,
+                    err=1e-6)
+    self.assertNear(np.log(9.),
+                    pkr_two.inner_results.inner_results.transformed_state,
+                    err=1e-6)
+
+  def test_nested_transform(self):
+    trans_kernel = tfp.mcmc.TransformedTransitionKernel(
+        inner_kernel=FakeInnerKernel(target_log_prob_fn=fake_target_log_prob),
+        bijector=tfb.Exp())
+    double_trans_kernel = tfp.mcmc.TransformedTransitionKernel(
+        inner_kernel=trans_kernel,
+        bijector=tfb.Identity())
+    step_kernel = tfp.mcmc.SimpleStepSizeAdaptation(
+        inner_kernel=double_trans_kernel,
+        num_adaptation_steps=9)
+    pkr_one, pkr_two = self.evaluate([
+        step_kernel.bootstrap_results(2.),
+        step_kernel.bootstrap_results(9.),
+    ])
+    self.assertNear(2., pkr_one.inner_results.transformed_state, err=1e-6)
```

Comment:
Did this actually work?  It seems like the transformed state should end being mapped through the composition of the transforming bijectors, rather than the last one?  Do we have a bug?

Reply:
Just checked to make sure and it does indeed pass. Here's the issue:

Without this PR such a composition wouldn't be possible as `TransformedTransitionKernel` doesn't have a `target_log_prob_fn` itself (it would've never been possible to nest two calls to `TransformedTransitionKernel` without error). Hence, currently, the `bootstrap_results` method just applies the "inverse" function of the single outermost bijector. Presumably, `one_step` has similar behavior as the "forward" function is similarly defined.

My question is if we want to support this kind of compositionality with multiple `TransformedTransitionKernel`s in this PR. If yes, I believe the easiest implementation would be to update the bijector in a similar manner to the `target_log_prob_fn`. I wouldn't know how the `log_det_jacobian` method naturally updates though. The alternative is to preface somewhere that this is the intended behaviour, and we're not meant to stack `TransformedTransitionkernel`s like this (it will override previous transformations).

Reply:
Final thoughts: intended behavior, write a test to make sure it is analogous to a bijector chain